### PR TITLE
Core: Use window.setTimeout & friends when provided

### PR DESCRIPTION
**Pull request by mzgol**
Fixes gh-2177

Alternative to #2391 & #2392. +40 bytes.

### CODE REVIEW

File: src/var/intervalClear.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+define(function() {
+	/* global clearInterval: false */
```
Comment (timmywil) [inline src/var/intervalClear.js]: Is this necessary since the globals are only removed for the built file?

Comment (mzgol) [inline src/var/intervalClear.js]: I removed `clearInterval` from JSHint globals but this is the one place where it has to appear.

File: src/var/intervalSet.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
+define(function() {
```
Comment (arthurvr) [inline src/var/intervalSet.js]: One day we should really decide if we're going to use those new spacing rules or not: `define( function() {`. We're inconsistent at the moment. It's the latest style guide but nobody on the team seems to like it.

File: src/var/intervalClear.js (hunk 1 above)
Comment (timmywil) [inline src/var/intervalClear.js]: Of course. I was reading it all backwards.

File: src/var/intervalSet.js (hunk 2 above)
Comment (mzgol) [inline src/var/intervalSet.js]: Yeah, I did it incorrectly, I'll fix. Thanks for noticing.

Comment (timmywil) [inline src/var/intervalSet.js]: shh

Comment (mzgol) [inline src/var/intervalSet.js]: You may notice I did it correctly in other places, though. ;) Here I just copied a template from another file, hence the mistake.

Once JSCS starts checking this, the problem will go away. ;)

Comment (mzgol) [inline src/var/intervalSet.js]: PR updated.

Comment (mzgol) [inline src/var/intervalSet.js]: OK, I can't change it because the regex for matching defines in `/var/*` modules expects no space: https://github.com/jquery/jquery/blob/8e111df641cca3e1b75b31a1971bfc89014b4ded/build/tasks/build.js#L14 (cc @timmywil)

Comment (markelog) [inline src/var/intervalSet.js]: > Once JSCS starts checking this, the problem will go away. ;)

Checking? Huh, this is past century talk! It will fix cases like that.

File: src/var/timeoutSet.js [hunk 3]
```diff
@@ -0,0 +1,4 @@
+define(function() {
+	/* global setTimeout: false */
+	return window.setTimeout || setTimeout;
```
Comment (markelog) [inline src/var/timeoutSet.js]: It looks weird without context, i would suggest add descriptive comments to each of those new files

Comment (mzgol) [inline src/var/timeoutSet.js]: This is a long comment, see https://github.com/mzgol/jquery/blob/7eef60f491be191e5844f3a3420bee238c7d73f3/src/intro.js#L19-L25. Should I repeat this comment in every of those 4 files?

File: src/var/intervalClear.js [hunk 4]
```diff
@@ -0,0 +1,4 @@
...
+	return window.clearInterval || clearInterval;
```
Comment (markelog) [inline src/var/intervalClear.js]: For which exactly `window.clearInterval` and just `clearInterval`? First is for jsdom and browser env, second for node?

File: src/var/timeoutSet.js (hunk 3 above)
Comment (markelog) [inline src/var/timeoutSet.js]: I'd say link should suffice

Comment (mzgol) [inline src/var/timeoutSet.js]: But link to what exactly? I linked to my first PR, this is the third where there is no central place to put it.

File: src/var/intervalClear.js (hunk 4 above)
Comment (mzgol) [inline src/var/intervalClear.js]: Yup.

**Comment by markelog** [comment]
I still question legitimacy of that ticket, we didn't receive any real issues about that and it seems only reason why we need to do that is

> However, some environments provide their own window.setTimeout etc.

Those environments should create browser-like conditions, not that other way around. If you think about it, it seems really weird, for us to do this, it is like for Firefox to change some their behaviour to keep consistency with the jsdom, which, actually should create Firefox like environment. Sounds like absurd, i bet no browser lib would do that.

In other words, jsdom should create hacks and not us, and again to emphasize - we didn't received real issues which should provoke this PR. 

Also, @domenic could you please check it out, since it seems, we doing that just for you.

Otherwise LGTM :-).

File: src/var/timeoutSet.js (hunk 3 above)
Comment (markelog) [inline src/var/timeoutSet.js]: Oh, then it would duplicate it everywhere :-(, would be really awkward though, or maybe just link to this pull would be enough?

**Comment by domenic** [comment]
There's no reason to do this. Scripts that run in jsdom can use `setTimeout` and friends just fine, with no prefix.

If you run into any other "do this to support jsdom" issues, let me know and we can fix jsdom.

**Comment by mzgol** [comment]
@markelog 
>> However, some environments provide their own window.setTimeout etc.
>
> Those environments should create browser-like conditions, not that other way around.

The thing is those environments *do* provide those conditions. The current situation is that we support both scenarios:

1. The global behaves like a browser window (web browsers, browserify): we return the jQuery object.
2. The global doesn't behave like a browser window: we then return a function to which you can provide your own `window` instance and it'll return the jQuery object using that `window`.

In the (2) case we're cheating a little as we generally take stuff off this `window` but with exception of `setTimeout` & friends which we always take from the global.

We could decide that this is an edge case that's not worth supporting but it will mean a little divergence from the (2) use case.

@domenic This mostly isn't about jsdom but about running in an environment where you can provide most needed stuff via your own `window` instance but without assuming you have anything on a global. Though if you were creating & closing (`window.close()`) a lot of windows in jsdom and using `setTimeout` etc. in them you'd leak memory, wouldn't you? `jsdom` has its own `setTimeout` etc. for a reason, doesn't it?

**Comment by mzgol** [comment]
> we didn't received any issues which should provoke this PR

We did, the exact one I mentioned in the description of this PR... i.e. #2177.

**Comment by domenic** [comment]
Oh, I don't really know anything about the use case of pasing in a window to a factory function. jsdom is geared around the use case of running jquery inside the jsdom environment itself, not that thing.

**Comment by markelog** [comment]
@mzgol well, in there, owner of that ticket tries to use Deferreds (in the environment that doesn't have setTimeout) which is better to do through https://github.com/jaubourg/jquery-deferred-for-node package or through custom build, so this is really out there from my perspective and doesn't worth 40 bytes and code comlication.

So i as i see it, since jsdom could work with jQuery just fine, the reason to do it is to be more consistent but a lot more hacky?

--
Thank you for the replay @domenic, you helped a lot.

**Comment by mzgol** [comment]
@domenic Hm. It was all added in 1f67d07c60c37e60052db37fc03d42af482c2d03. You even posted some comments under the ticket. ;)

**Comment by mzgol** [comment]
Well, if we just want to assume that `window` is available then the whole factory function with all this complicated logic when to return the object and when the function accepting `window` as a parameter is unnecessary... So how did we end up here? :P @timmywil?

**Comment by mzgol** [comment]
Also, if we want to go full-in in requiring a proper window to be present
then I'd prefer to remove the JSHint exception for setTimeout etc. and just
use window.setTimeout without a fallback to the global one.

-- 
Michał Gołębiowski

**Comment by timmywil** [comment]
Actually, I'd be fine with requiring a proper window and moving to window.setTimeout everywhere. The problems that would cause in Node would be mitigated by using tools like jsdom along with jQuery, which is probably the only way jQuery should be used in Node anyway.

Side note: we probably shouldn't recommend https://github.com/jaubourg/jquery-deferred-for-node after 3.0 comes out. It hasn't received updates in a long time and does not support Promises/A+.

**Comment by mzgol** [comment]
@timmywil @markelog OK, I created yet another attempt: #2397.

**Comment by markelog** [comment]
> Side note: we probably shouldn't recommend https://github.com/jaubourg/jquery-deferred-for-node after 

Let's ask @jaubourg, i thought he is still commited to supporting this module

**Comment by markelog** [comment]
> I'd be fine with requiring a proper window and moving to window.setTimeout everywhere

Why we need to window "prefix" for global vars?

File: src/var/intervalSet.js (hunk 2 above)
Comment (markelog) [inline src/var/intervalSet.js]: @mzgol btw, coud you create a ticket about that?

**Comment by mzgol** [comment]
> Why we need a window "prefix" for global vars?

For the same reason we prefix `getComputedStyle`, `document`, `document.documentElement` etc. - so that we're consistent in taking them from the provided `window` instance instead of the global. If we e.g. assumed a global `getComputedStyle`, jsdom wouldn't work. It would work without scoping `setTimeout` but there are other use cases. I think we should be consistent.

**Comment by dmethvin** [comment]
It took me a while to catch up. In this particular case on #2177 it seems like the methods would normally be on `global` but are only on `window` for this unusual Firefox env, so we reference them from there?

File: src/var/intervalSet.js (hunk 2 above)
Comment (mzgol) [inline src/var/intervalSet.js]: @markelog #2399

**Comment by mzgol** [comment]
@dmethvin This comment of mine is a good summary: https://github.com/jquery/jquery/pull/2394#issuecomment-112192426

We take most globals from `window` but we take `setTimeout` etc. from the global which works for jsdom but not necessarily for other envs and is inconsistent. This PR tries to support both but the simpler one (the 4th my PR trying to fix this issue ;)) that just takes them from `window` is at #2397.

**Comment by markelog** [comment]
Now i remember, okay, i agree, but i still would just write `window.setTimeout`

**Comment by mzgol** [comment]
Closing in favor of #2397.
